File: LeetCode/106-construct-binary-tree-from-inorder-and-postorder-traversal.py
###############################################################################################
# simple recursion
###########
# Time Complexity: O(nlogn), n is the number of nodes in the tree
# Space Complexity: O(logn)
###############################################################################################
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        def dfs(ino, poo):
            if not ino or not poo:
                return None
            idx = ino.index(poo[-1])
            root = TreeNode(poo[-1], None, None)
            root.left = dfs(ino[0:idx], poo[0:idx])
            root.right = dfs(ino[idx+1:], poo[idx:-1])
            return root
        return dfs(inorder, postorder)

File: LeetCode/106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// not hard
////////////////
// Time Complexity: O(n)
// Space Complexity: O(n)
///////////////////////////////////////////////////////////////////////////////////////////////
class Solution {
public:
    TreeNode* buildTreeHelper(vector<int>& inorder, vector<int>& postorder, int inorderStart, int inorderEnd, int postorderStart, int postorderEnd, unordered_map<int, int>& index) {
        if (inorderStart > inorderEnd || postorderStart > postorderEnd) return nullptr;

        // get cur root val
        int rootVal = postorder[postorderEnd];
        TreeNode* root = new TreeNode(rootVal);

        // get it's idx in inorder
        int inorderRootIndex = index[rootVal];

        int leftSubtreeSize = inorderRootIndex - inorderStart;
        root->left = buildTreeHelper(inorder, postorder, inorderStart, inorderRootIndex - 1, postorderStart, postorderStart + leftSubtreeSize - 1, index);
        root->right = buildTreeHelper(inorder, postorder, inorderRootIndex + 1, inorderEnd, postorderStart + leftSubtreeSize, postorderEnd - 1, index);

        return root;
    }
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        unordered_map<int, int> index;
        for (int i = 0; i < inorder.size(); i++) index[inorder[i]] = i;
        return buildTreeHelper(inorder, postorder, 0, inorder.size() - 1, 0, postorder.size() - 1, index);
    }
};